README

*****************************************************************************
*****************************************************************************
*                                                                           *
*                           Preliminaries                                   *
*                                                                           *
*****************************************************************************
*****************************************************************************
Configuration and build for 5.0.2 were tested with:
aclocal,automake  1.9.4 and 1.9.6
autoheader,autoconf 2.59, 2.60
gcc 3.4.4 and 3.4.6 (see also README.compilers)

RedHat    Linux 2.6.9-42.0.8.ELsmp #1 SMP Tue Jan 30 12:33:47 EST 2007 i686 
Slackware Linux 2.4.33.3 #1 2.4.33.3 #1 Fri Sep 1 01:48:52 CDT 2006 i686 



*****************************************************************************
*****************************************************************************
*                                                                           *
*                           Configuring                                     *
*                                                                           *
*****************************************************************************
*****************************************************************************
Run 
	./configure 
    or
        ./configure --prefix=<alt-to-/usr/local-for-install>
    or
	./configure  --help 
           to see options.

If that doesn't work or if you have been hacking on the
storage manager and if you have the autotools installed, 
you can run

	./bootstrap

followed by configure.

-----------------------------------------------------------------------------
NOTE: re: configure options : run
	./configure --help 
to find out what SHORE-specific build configure offers. 

--enable-lp64  : this is the default.  This option exists only for the
                 purpose of enabling building on a 32-bit platform, which
				 is NOT SUPPORTED YET.

--enable-checkrc: when enabled, code is included to make sure that 
                 a w_rc_t has been checked for is_error() before it is
				 destructred.

--enable-trace  : when enabled, code is included to print tracing info
                 based on the value of the DEBUG_FLAGS environment variable

--enable-dbgsymbols  : when enabled, turns on -g, but can be overridden by
                 compiler optimization options, depending  compiler and
				 on debugging level used

--enable-explicit  : when enabled, compiles with explicit templates if the
                 compiler supports it. NOT SUPPORTED NOW



If you are planning to modify the storage manager, you'll probably
want to configure a debug version.

--with-debug-level0 :  default if no other level is given.
                       Max optimization,   inlining
					       no debugger symbols
						   defines W_DEBUG_LEVEL=0, enables w_assert0
						   -UDEBUG, -DNDEBUG
					   This version includes little or no self-checks (asserts), so is
					   not good for "production" unless you are sure about your code.
					   Useful for performance testing, if you're trusting.

--with-debug-level1 :  Optimization -O2,  inlining
					       includes debugger symbols
                           defines W_DEBUG_LEVEL=1, enables w_assert1
						   -UDEBUG, -DNDEBUG
					   This version is probably a good "standard" version.

--with-debug-level2 :  No optimization,  no inlining
					       includes debugger symbols
                           defines W_DEBUG_LEVEL=2, enables w_assert1, w_assert2
						   -UDEBUG, -DNDEBUG
					   This version changes the timing quite a bit but is
					   good for debugging functionality that isn't timing-dependent.

--with-debug-level3 :  No optimization,  no inlining
					       includes debugger symbols
                           defines W_DEBUG_LEVEL=3, enables w_assert1, w_assert2, w_assert3
						   -DDEBUG, -UNDEBUG 
					   This version changes the timing quite a bit but is
					   good for debugging functionality that isn't timing-dependent,
					   and includes a lot more self-checking than level 2.
                     
--with-debug-levelN with N>3 : not supported yet. 
                           Any code you see with
							   w_assert<N> for N>3, or 
							   relating to W_DEBUG_LEVELs greater than 3 
						   is  experimental or code that is left over from
						   pre-multithreaded days and is not tested yet. 
						   Or it is disabled for another reason.


When in doubt, you can see the flags used by the last make in the file ./makeflags.

-----------------------------------------------------------------------------
NOTE: If you want to use a compiler other than that which ./configure
finds, do this:
	setenv CXX <path-to-compiler>
then 
	./configure

or
	./configure CXX=<path-to-compiler>
-----------------------------------------------------------------------------
NOTE: If you don't seem to be getting dependency tracking and you want
it to be enabled, add --enable-dependency-tracking to your ./configure
command.

-----------------------------------------------------------------------------
   

See also README.cppmacros




*****************************************************************************
*****************************************************************************
*                                                                           *
*                           Building                                        *
*                                                                           *
*****************************************************************************
*****************************************************************************
After running configure,  run

	make 

to make the libraries.
-----------------------------------------------------------------------------

NOTE: the documents are distributed in PDF form and cannot be made.
They can only be installed with 'make install'.



*****************************************************************************
*****************************************************************************
*                                                                           *
*                 Checking the built libraries                              *
*                                                                           *
*****************************************************************************
*****************************************************************************

Run
	make check 
to build and run the test programs via scripts in each library.

-----------------------------------------------------------------------------
--->NOTE: To make smsh, the storage manager testing shell, you must edit 

           Makefile.local

so that make can find your local TCL installation, if it's not in a standard
place where it will be found (/usr/include/tcl.h, /usr/lib/libtcl.so, etc).
-----------------------------------------------------------------------------
The 'testall' script in smsh runs a very long set of tests, so 
it is not the default test for smsh.
-----------------------------------------------------------------------------



*****************************************************************************
*****************************************************************************
*                                                                           *
*                           Installing                                      *
*                                                                           *
*****************************************************************************
*****************************************************************************

-----------------------------------------------------------------------------
Run
	make install 

to install 
	diskrw executable in <bindir> [default: <exec-prefix>/bin]
	the header files  in <includedir> [default: <prefix>/include]
	the libraries in <libdir> [default: <exec-prefix>/lib]
	the documents in <docdir> [default: <prefix>/share/doc/shore-storage-manager]

To change the generic prefix, use the configure option:
	configure --prefix=<new prefix path>
To change the specific target directories, use one or more of the ./configure option:
	configure --bindir=<new bindir path>
	configure --includedir=<new include path>
	configure --libdir=<new libdir path>
	configure --docdir=<new docdir path>

    NOTE: despite what newer versions of configure say about pdfdir, htmldir, 
          psdir, you cannot use these options with this distribution.

-----------------------------------------------------------------------------


ChangeLog
Distributions of Shore Storage Manager 

5.0 : Thu Jun 7 15:25 2007
      from starting point being the 4th interim release ...
      1) removed the Imake-based configuration and replaced it with autoconf
      2) changes ssh to smsh

5.0.1: Sat Aug 18 15:37:16 CDT 2007
      1) added the smsh scripts to the distr


Branch mt-branch: Mon Jun 22 20:03:29 CDT 2009
{

	Makes with gcc 4.3.2 

	Added debuglevels to config but they are not really tested yet.
	Mostly I made with -g and -UDEBUG -UW_DEBUG -DNDEBUG, which included
	w_assert1 but not the others.


	I got through 'make check' in fc/, sthread/, common (except stats, which
	have to be thoroughly rethought), sm/tests (file_scan)
	and smsh with one script: vol.1

	Known problems: the whole smsh thing has to be studied.
	The whole debugging apparatus has to be studied and redone.
	Much has to be ripped out and cleaned up.
	---------------


	I cleaned up the largefile-awareness a tiny bit. The ASSUMPTION now
	is that all platforms handle largefile-awareness, even if they are
	not ARCH_LP64.

	I'm starting to rip out anything for other platforms besides
	solaris/sparcv9 and linux/i686 -- all that OSF1 and MACOS and VC++
	stuff will go, and already it has started its exit.

	---------------

	The following naming  changes are made here:

	fc/getopt.[ch] --> fc/w_getopt.[ch]
	sthread/sthread.h ripped out the block_alloc and atomic_stack and put
	   them into fc/block_alloc.h and fc/atomic_container.h

	The following algorithmic changes are made here (this note taken from
	mail sent to Ippokratis):


	1) log : I take it mcs_log is not in use, since you told me
	you were using ringbuf_log. So I took out mcs log, but there was an oddity
	in that the sm was using mcs_log::new_log_m, so I changed the name
	of that function to srv_log::new_log_m and was able to get rid of the
	mcs_log.

	2) It was possible to have one thread shutting down the sm before
	another thread, namely the log's flush daemon, 
	tried to dereference its _log ptr, which, 
	owing to its being destructed, no longer could be dereferenced to execute
	_flush().  I added a shutdown method on log_m and ringbuf_log; the
	sm calls shutdown to to whatever needs to be done before the destructor
	can safely be run, namely, the first few lines of what was in the 
	destructor.

	3) recovery could get into an infinite loop in vol.cpp, recover_pages_in_ext:

	replace this:
	<       for(int i=mypmap.first_set(0); i < mypmap.count(); i=mypmap.first_set(i+1)) 
	with this:
	>       for(int i=mypmap.first_set(0);
	>               (i>=0)&& (i < mypmap.count()); i=mypmap.first_set(i+1)
	>       ) 


	4) io_m::_alloc_pages
	 I had to turn off the caching of extents (there was an #if 1/#endif
	 around it, so that is convenient), since it was breaking some simple
	 sm tests.  Here's my note to myself about it:

	 > // TODO: NANCY : bug here. If we turn on the caching above (#if 0/1)
	> // we crap out here because last_page is looking for the last allocated
	> // page (not the last page - reserved or allocated - because we aren't
	> // passing in a Boolean ptr.  In the original code, we wouldn't be
	> // doing this unless we KNEW that we HAD to allocate more extents because
	> // we would already have allocated all from existing extents. The
	> // caching behavior changes that into a hint, so we can get here
	> // having extents w/ no allocated pages in them.
	> // I don't want to get side-tracked fixing this yet; I'm in the middle
	> // of too many things.   Let that be a later project. For now,
	> // just turn off the caching.
	>

	The caching was a performance improvement but it wasn't thoroughly tested,
	so for now I'm turning it off and will get back to it later. 
	Food for thought. Maybe you can get it working with the caching in place before
	I get to it.

	5) chkpt.cpp: I move the xlist_mutex_acquire outside the loop so it
	doesn't doubly-acqure the mutex.  (The pthreads
	mutex lock will block you forever if you already hold the lock).
	(I'm just trying to build with pthreads mutexes initially.)

}



src/fc/w.h
/*<std-header orig-src='shore' incl-file-exclusion='W_H'>

 $Id: w.h,v 1.19.2.5 2010/03/19 22:17:19 nhall Exp $

SHORE -- Scalable Heterogeneous Object REpository

Copyright (c) 1994-99 Computer Sciences Department, University of
                      Wisconsin -- Madison
All Rights Reserved.

Permission to use, copy, modify and distribute this software and its
documentation is hereby granted, provided that both the copyright
notice and this permission notice appear in all copies of the
software, derivative works or modified versions, and any portions
thereof, and that both notices appear in supporting documentation.

THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
"AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.

This software was developed with support by the Advanced Research
Project Agency, ARPA order number 018 (formerly 8230), monitored by
the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
Further funding for this work was provided by DARPA through
Rome Research Laboratory Contract No. F30602-97-2-0247.

*/


/** \mainpage SHORE Storage Manager: The Multi-Threaded Version 
 * \section Brief Description
 *
 * This is an experiment test-bed library for use by researchers who wish to
 * write multi-threaded software that manages persistent data.   
 *
 * This storage engine provides the following capabilities:
 *  - transactions with ACID properties, with ARIES-based logging and recovery,
 *  primitives for partial rollback,
 *  transaction chaining, and early lock release,
 *  - prepared-transaction support for two-phased commit,
 *  - persistent storage structures : 
 *  B+ tree indexes, R* trees (spatial indexes), and files of untyped records,
 *  - fine-grained locking for records and B+ tree indexes with deadlock detection,
 *  optional lock escalation and optional coarse-grained locking, 
 *  - in-memory buffer management with optional prefetching, 
 *  - extensible statistics-gathering, option-processing, and error-handling 
 *  facilities.
 *
 * This software runs on Pthreads, thereby providing its client software
 * (e.g., a database server) multi-threading
 * capabilities and resulting scalability from modern SMP and NUMA 
 * architectures, and has been used on Linux/x86-64 and Solaris/Niagara
 * architectures.
 *
 * \section Background
 *
 * The SHORE (Scalable Heterogeneous Object REpository) project
 * at the University of Wisconsin - Madison Department of Computer Sciences
 * produced the first release
 * of this storage manager as part of the full SHORE release in 1996.
 * The storage manager portion of the SHORE project was used by 
 * other projects at the UW and elsewhere, and was intermittently
 * maintained through 2008.
 *
 * The SHORE Storage Manager was originally developed on single-cpu Unix-based systems,
 * providing support for "value-added" cooperating peer servers, one of which was the
 * SHORE Value-Added Server (http://www.cs.wisc.edu/shore), and another of which was 
 * Paradise (http://www.cs.wisc.edu/paradise) at the University of Wisconsin;
 * TIMBER (http://www.eecs.umich.edu/db/timber) and 
 * Persicope (http://www.eecs.umich.edu/persiscope) at the University of Michigan,
 * and
 * PREDATOR (http://www.distlab.dk/predator) at Cornell.
 * (The storage manager was also used for innumerable published studies.)
 *
 * The storage manager had its own "green threads" and communications
 * layers, and until 2007, its code structure, nomenclature, and contents reflected
 * its SHORE roots.
 *
 * In 2007, the Data Intensive Applications and Systems Labaratory (DIAS)
 * at Ecole Polytechnique Federale de Lausanne
 * began work on a port of release 5.0.1 of the storage manager to Pthreads,
 * and developed more scalable synchronization primitives, identified
 * bottlenecks in the storage manager, and improved the scalability of 
 * the code.
 * This work was on a Solaris/Niagara platform and was released as Shore-MT
 * http://diaswww.epfl.ch/shore-mt).
 * It was a partial port of the storage manager and did not include documentation.
 * Projects using Shore-MT include
 * StagedDB/CMP (http://www.cs.cmu.edu/~stageddb/),
 * Lachesis (http://www.vldb.org/conf/2003/papers/S21P03.pdf), 
 * and
 * DORA (http://www.cs.cmu.edu/~ipandis/resources/CMU-CS-10-101.pdf)
 *
 * In 2009, the University of Wisconsin - Madison took the first Shore-MT
 * release and ported the remaining code to Pthreads.
 * This work as done on a Red Hat Linux/x86-64 platform.
 * This release is the result of that work, and includes this documentation,
 * bug fixes, and supporting test code.   Some of the scalability changes
 * of the DIAS
 * release has been removed in order to correct bugs, with the hope that 
 * further work will improve scalability of the completed port.
 *
 * \section Copyrights
 * Most of this library and documentation is subject one or both of the following copyrights:
 *
 *  -       \b SHORE -- \b Scalable \b Heterogeneous \b Object \b REpository
 *
 *  Copyright (c) 1994-2007 Computer Sciences Department, University of
 *                      Wisconsin -- Madison
 *                      All Rights Reserved.
 *
 *  Permission to use, copy, modify and distribute this software and its
 *  documentation is hereby granted, provided that both the copyright
 *  notice and this permission notice appear in all copies of the
 *  software, derivative works or modified versions, and any portions
 *  thereof, and that both notices appear in supporting documentation.
 *
 *  THE AUTHORS AND THE COMPUTER SCIENCES DEPARTMENT OF THE UNIVERSITY
 *  OF WISCONSIN - MADISON ALLOW FREE USE OF THIS SOFTWARE IN ITS
 *  "AS IS" CONDITION, AND THEY DISCLAIM ANY LIABILITY OF ANY KIND
 *  FOR ANY DAMAGES WHATSOEVER RESULTING FROM THE USE OF THIS SOFTWARE.
 *
 *  This software was developed with support by the Advanced Research
 *  Project Agency, ARPA order number 018 (formerly 8230), monitored by
 *  the U.S. Army Research Laboratory under contract DAAB07-91-C-Q518.
 *  Further funding for this work was provided by DARPA through
 *  Rome Research Laboratory Contract No. F30602-97-2-0247.
 *
 * -   \b Shore-MT -- \b Multi-threaded \b port \b of \b the \b SHORE \b Storage \b Manager
 *  
 *                     Copyright (c) 2007-2009
 *        Data Intensive Applications and Systems Labaratory (DIAS)
 *                 Ecole Polytechnique Federale de Lausanne
 *  
 *                       All Rights Reserved.
 *  
 *  Permission to use, copy, modify and distribute this software and
 *  its documentation is hereby granted, provided that both the
 *  copyright notice and this permission notice appear in all copies of
 *  the software, derivative works or modified versions, and any
 *  portions thereof, and that both notices appear in supporting
 *  documentation.
 *  
 *  This code is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
 *  DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
 *  RESULTING FROM THE USE OF THIS SOFTWARE.
 *
 * \section START Getting Started With the Shore Storage Manager
 * A good place to start is with \ref SSMAPI.
 */

/**\addtogroup OPT
 *\section CONFIGOPT Configuration Options
 * \verbatim
SHORE-specific Features:
  --enable-lp64         default:yes     Compile to use LP 64 data model
                                        No other data model is supported yet.
                                        But we hope some day to port back to LP32.
  --enable-checkrc      default:no      Generate (expensive) code to verify return-code checking
                                        If a w_rc_t is set but not checked with
                                        method is_error(), upon destruction the
                                        w_rc_t will print a message to the effect
                                        "error not checked".
  --enable-trace        default:no      Include tracing code
                                        Run-time costly.  Good for debugging
                                        problems that are not timing-dependent.
                                        Use with DEBUG_FLAGS and DEBUG_FILE
                                        environment variables.  See \ref SSMTRACE.
  --enable-dbgsymbols   default:no      Turn on debugger symbols
                                        Use this to override what a given
                                        debugging level will normally do.
  --enable-explicit     default:no      Compile with explicit templates
                                        NOT TESTED. \todo explicit templates

  --enable-valgrind     default:no      Enable running valgrind run-time behavior
                                        Includes some code for valgrind.
  --enable-purify       default:no      Enable build of <prog>.pure
  --enable-quantify     default:no      Enable build of <prog>.quant
  --enable-purecov      default:no      Enable build of <prog>.purecov

SHORE-specific Optional Packages:
  --with-hugetlbfs        Use the hugetlbfs for the buffer pool.
                          Depending on the target architecture, this might
                          be useful.  If you use it, you will need to set
                          a path for your hugetlbfs in config/shore.def.
                          The default is :
                          #define HUGETLBFS_PATH "/mnt/huge/SSM-BUFPOOL"
  --without-mmap          Do not use mmap for the buffer pool. Trumps
                          hugetlbfs option.
  --with-debug-level1     Include level 1 debug code, optimize.
                          This includes code in w_assert1 and 
                          #if W_DEBUG_LEVEL > 0 /#endif pairs and 
                          #if W_DEBUG_LEVEL >= 1 /#endif pairs and  and
                          W_IFDEBUG1
  --with-debug-level2     Include level 2 debug code, no optimize.
                          Equivalent to debug level 1 PLUS
                          code in w_assert2 and
                          #if W_DEBUG_LEVEL > 1 /#endif pairs and
                          #if W_DEBUG_LEVEL >= 2 /#endif pairs and
                          W_IFDEBUG2
  --with-debug-level3     Include level 3 debug code, no optimize.
                          Equivalent to debug level 2 PLUS
                          includes code in w_assert3  and
                          #if W_DEBUG_LEVEL > 2 /#endif pairs and
                          #if W_DEBUG_LEVEL >= 3 /#endif pairs and
                          W_IFDEBUG3
 \endverbatim
 */

/**\addtogroup OPT
 * \section SHOREDEFOPT Build options in config/shore.def
 * In this section we describe C preprocessor macros defined (or not) in
 * config/shore.def.
 *
 * -HUGETLBFS_PATH See --with-hugetlbfs in \ref CONFIGOPT
 *
 * -USE_SSMTEST Define this if you want to include crash test hooks in your
 *
 * -DONT_TRUST_PAGE_LSN -- this is a deprecated option.
 * executable.  (For use with ssh, really).
 * A release executable should not be built with this unless you
 * are a maintainer and want to test it. It is not documented.
 *
 * -VOL_LOCK_IS_RW  Undefine this if you want to restore the non-r-w mutex for
 * the volume (vol_t) structure.
 * The volume mutex was replaced with a read-write lock because 
 * statistics showed that the vast majority of the times we grab the mutex, it
 * is for read-operations. 
 * Having changed this to a read-write lock, we now see that
 * <10% of the time does the read-lock request have to wait,
 * and roughly 40% of the time the write-lock request has to wait.  
 * As this code is still somewhat experimental, the option to turn it off
 * remains, for the time being.
 */

/**\defgroup SSMAPI SHORE Storage Manager Application Programming Interface (SSM API)
 */
/**\defgroup MISC Miscellaneous
 */
/**\defgroup Macros Significant macros 
 * \ingroup MISC
 *
 * Not all macros are included here.
 * */
/**\defgroup UNUSED Unused code 
 * \ingroup MISC
 */

#ifndef W_H
#define W_H

#include "w_defines.h"

/*  -- do not edit anything above this line --   </std-header>*/

#include <w_base.h>
#include <w_minmax.h>
#include <w_list.h>
#include <w_hash.h>

#ifdef W_INCL_BITMAP
#include <w_bitmap.h>
#endif

/*<std-footer incl-file-exclusion='W_H'>  -- do not edit anything below this line -- */

#endif          /*</std-footer>*/
